master/internal/command/command.go

```diff
@@ -146,357 +139,326 @@ func remakeCommandsByType(
+	cmd := &command{
+		db: cs.db,
+		rm: cs.rm,
 
-	registeredTime time.Time
-	taskID         model.TaskID
-	taskType       model.TaskType
-	jobType        model.JobType
-	jobID          model.JobID
-	allocationID   model.AllocationID
-	lastState      task.AllocationState
-	exitStatus     *task.AllocationExited
-	restored       bool
+		GenericCommandSpec: *spec,
 
-	contextDirectory []byte // Don't rely on this being set outsides of PreStart non restore case.
+		taskID:           taskID,
+		taskType:         taskType,
+		jobType:          jobType,
+		jobID:            jobID,
+		contextDirectory: contextDirectory,
+		logCtx: logger.Context{
+			"job-id":    jobID,
+			"task-id":   taskID,
+			"task-type": taskType,
+		},
+		syslog: logrus.WithFields(logrus.Fields{
+			"component": "command",
+			"task-id":   taskID,
+		}),
+	}
+	// Add it to the registry.
+	cs.commands[cmd.taskID] = cmd
 
-	logCtx logger.Context
-	syslog *logrus.Entry
+	return cmd, cmd.startCmd(ctx)
 }
 
-// Receive implements the actor.Actor interface.
-func (c *command) Receive(ctx *actor.Context) error {
+// start starts the command & its respective allocation. Once started, it persists to the db.
+func (c *command) startCmd(ctx context.Context) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	switch msg := ctx.Message().(type) {
-	case actor.PreStart:
-		priorityChange := func(priority int) error {
-			return c.SetPriority(priority, false)
-		}
-		if err := tasklist.GroupPriorityChangeRegistry.Add(c.jobID, priorityChange); err != nil {
+	if !c.restored {
+		if err := db.Bun().RunInTx(ctx, nil, c.registerJobAndTask); err != nil {
 			return err
 		}
-		ctx.AddLabels(c.logCtx)
-		c.allocationID = model.AllocationID(fmt.Sprintf("%s.%d", c.taskID, 1))
-		if !c.restored {
-			// TODO all this stuff should be in transactions.
-			c.registeredTime = ctx.Self().RegisteredTime().Truncate(time.Millisecond)
-			if err := c.db.AddJob(&model.Job{
-				JobID:   c.jobID,
-				JobType: c.jobType,
-				OwnerID: &c.Base.Owner.ID,
-			}); err != nil {
-				return errors.Wrapf(err, "persisting job %v", c.taskID)
-			}
-
-			if err := c.db.AddTask(&model.Task{
-				TaskID:     c.taskID,
-				TaskType:   c.taskType,
-				StartTime:  c.registeredTime,
-				JobID:      &c.jobID,
-				LogVersion: model.CurrentTaskLogVersion,
-			}); err != nil {
-				return errors.Wrapf(err, "persisting task %v", c.taskID)
-			}
-
-			if err := c.persistAndEvictContextDirectoryFromMemory(); err != nil {
-				return err
-			}
-		}
+	}
 
-		priority := c.Config.Resources.Priority
-		if priority != nil {
-			if err := c.setPriority(*priority, true); err != nil {
-				return errors.Wrapf(err, "setting priority of task %v", c.taskID)
-			}
-		}
+	priorityChange := func(priority int) error {
+		return c.setNTSCPriority(priority, false)
+	}
+	if err := tasklist.GroupPriorityChangeRegistry.Add(c.jobID, priorityChange); err != nil {
+		return err
+	}
+
+	c.allocationID = model.AllocationID(fmt.Sprintf("%s.%d", c.taskID, 1))
```

Comment:
did we change the order of this code (this runs after the `c.restored` block) for any reason? if not, put it back to minimize risk?